prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,478 @@
+/* SPDX-License-Identifier: BSD-2-Clause OR GPL-2.0-only */
+/* SPDX-FileCopyrightText: Copyright (c) 2013-2015 Intel Corporation, Inc.  All rights reserved. */
+/* SPDX-FileCopyrightText: Copyright Amazon.com, Inc. or its affiliates. All rights reserved. */
+
+#include <errno.h>
+#include <string.h>
+#include "config.h"
+#include <ofi_mem.h>
+#include "dgram/efa_dgram_ep.h"
+#include "efa.h"
+#include "efa_cq.h"
+#include "efa_av.h"
+#include <infiniband/verbs.h>
+
+
+static inline uint64_t efa_cq_opcode_to_fi_flags(enum ibv_wc_opcode opcode) {
+	switch (opcode) {
+	case IBV_WC_SEND:
+		return FI_SEND | FI_MSG;
+	case IBV_WC_RECV:
+		return FI_RECV | FI_MSG;
+	case IBV_WC_RDMA_WRITE:
+		return FI_RMA | FI_WRITE;
+	case IBV_WC_RECV_RDMA_WITH_IMM:
+		return FI_REMOTE_CQ_DATA | FI_RMA | FI_REMOTE_WRITE;
+	case IBV_WC_RDMA_READ:
+		return FI_RMA | FI_READ;
+	default:
+		assert(0);
+		return 0;
+	}
+}
+
+static void efa_cq_construct_cq_entry(struct ibv_cq_ex *ibv_cqx, struct fi_cq_tagged_entry *entry)
+{
+	entry->op_context = (void *)ibv_cqx->wr_id;
+	entry->flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cqx));
+	entry->len = ibv_wc_read_byte_len(ibv_cqx);
+	entry->buf = NULL;
+	entry->data = 0;
+	entry->tag = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cqx) & IBV_WC_WITH_IMM) {
+		entry->flags |= FI_REMOTE_CQ_DATA;
+		entry->data = ibv_wc_read_imm_data(ibv_cqx);
+	}
+}
+
+/**
+ * @brief handle the situation that a TX/RX operation encountered error
+ *
+ * This function does the following to handle error:
+ *
+ * 1. write an error cq entry for the operation, if writing
+ *    CQ error entry failed, it will write eq entry.
+ *
+ * 2. increase error counter.
+ *
+ * 3. print warning message with self and peer's raw address
+ *
+ * @param[in]	base_ep     efa_base_ep
+ * @param[in]	ibv_cq_ex   extended ibv cq
+ * @param[in]	util_cq     tx/rx cq to write error to
+ * @param[in]	err         positive libfabric error code
+ * @param[in]	prov_errno  positive EFA provider specific error code
+ */
+static void efa_cq_handle_error(struct efa_base_ep *base_ep,
+				struct ibv_cq_ex *ibv_cq_ex,
+				struct util_cq *util_cq, int err,
+				int prov_errno)
+{
+	struct fi_cq_err_entry err_entry;
+	fi_addr_t addr;
+	int write_cq_err;
+
+	memset(&err_entry, 0, sizeof(err_entry));
+
+	err_entry.err = err;
+	err_entry.prov_errno = prov_errno;
+	err_entry.op_context = (void *)(uintptr_t)ibv_cq_ex->wr_id;
+	err_entry.flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cq_ex));
+	err_entry.buf = NULL;
+	err_entry.data = 0;
+	err_entry.tag = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cq_ex) & IBV_WC_WITH_IMM) {
+		err_entry.flags |= FI_REMOTE_CQ_DATA;
+		err_entry.data = ibv_wc_read_imm_data(ibv_cq_ex);
+	}
+
+	// TODO: get correct peer addr for tx
+	addr = efa_av_reverse_lookup(base_ep->av,
```

Comment:
Before we get it correct, we should make this addr retrieve only for rx operation. For tx operation, make it temporarily as `FI_ADDR_NOTAVAIL`